### PR TITLE
Enable Deposits Support

### PR DESCRIPTION
**Pull request by peterwilsoncc**
Supersedes #3855
Related WOODEP-1
Related https://github.com/woocommerce/woocommerce-deposits/issues/715 (specific task)
Related https://github.com/woocommerce/woocommerce-deposits/issues/375 (project)

## Changes proposed in this Pull Request:

Introduces support of the `deposits` feature for processing scheduled and subsequent orders in WooCommerce Deposits. This is to allow merchants using the Deposits extension to be paid automatically on the payment plan schedule and as the merchant completes orders with a fixed or percentage deposit.

Introduces `WC_Stripe_Deposits_Trait` for supporting the feature.

Initilization of the feature requires:
* the presence of the `wc_deposits_init()` function (available as the Deposits extension is included)
* the gateway supports `tokenization`

Similar to Pre-Orders, the token/intent is stored against the order rather than as a user payment method. The intent is detached once the order is fully paid to prevent live intents from being stored in the database longer than they are required. The intent details are retained after the final payment to allow for refunds.

This does not include any UI changes so testing on mobile is not required.

> [!Note]
>
> PHPUnit tests have been started but a currently broken.
>
> I'm going to need some assistance about how to test with Deposits. I was going to base them on the pre-orders tests but I can't see what I need to mock differently for Desposits.

## Testing instructions

1. Enable Deposits using the branch https://github.com/woocommerce/woocommerce-deposits/pull/720
2. Create a Deposits payment plan product:
   * Name the Product: `Deposits 2 day payment plan` -- I'll use this name below
   * Account for the minimum allowed charge for each payment
   * Two day payment plan
   * Payment plan is required
   * 50% at checkout, 50% 1 day later
3. Create a Deposits fixed amount product
   * Name the product: `$2 Deposits product ($1 deposit)` -- I'll use this name below
   * You may need to modify the amounts to account for minimum payment amounts
   * Deposit is required
   * Total cost: $2
   * Deposit amount: $1
4. Create a Deposits percentage amount product
   * Name the product: `$2 Deposits product (50% deposit)` -- I'll use this name below
   * You may need to modify the amounts to account for minimum payment amounts
   * Deposit is required
   * Total cost: $2
   * Deposit amount: 50%
5. Purchase the product `Deposits 2 day payment plan`
6. Ensure the 50% upfront charge is processed by Stripe successfully
7. In the database, modify the scheduled payment date to bring it forward (I usually set it to a few seconds after the intial purchase). The order can be found in the `wp_wc_orders` table, the charge date to modify is the `date_created_gmt` column
8. Run the WP CLI command `wp eval "do_action( 'woocommerce_invoice_scheduled_orders' );"` to emulate the cron job firing.
9. Wait two or three minutes for the Action Scheduler tasks to complete, the second order will show as processing or completed once this is done. If the second order status shows as pending payment, this step has failed.
10. View the stripe logs in the dashboard.stripe.com
11. Ensure that GET requests are selected in the Methods filter
12. Ensure that the payment intent `detach` endpoint has been hit
13. Purchase the `$2 Deposits product ($1 deposit)` product
14. Ensure the $1 upfront charge is processed by Stripe successfully
15. View the order in the WP Dashboard
16. Click the "Charge remaining amount" button below the line item
17. You will be redirected to the new order, ensure this:
   * is showing as paid
   * is showing as charging the customer $1
19. View the stripe logs in dashboard.stripe.com
20. Ensure that GET requests are selected in the Methods filter
21. Ensure the payment intent `detach` endpoint has been hit
22. Purchase the `$2 Deposits product (50% deposit)` product
23. Ensure the $1 upfront charge is processed by Stripe successfully
24. View the order in the WP Dashboard
25. Click the "Charge remaining amount" button below the line item
17. You will be redirected to the new order, ensure this:
   * is showing as paid
   * is showing as charging the customer $1
26. View the stripe logs in dashboard.stripe.com
27. Ensure that GET requests are selected in the Methods filter
28. Ensure the payment intent `detach` endpoint has been hit

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Tested on mobile (or does not apply)

### Changelog entry

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Add - Introduce support for automatic payment processing in WooCommerce Deposits</summary>

#### Comment 

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: tests/phpunit/payment-methods/test-class-wc-stripe-upe-payment-gateway.php [hunk 1]
```diff
@@ -2240,6 +2242,64 @@ public function test_pre_order_without_payment_uses_setup_intents() {
 		$this->assertTrue( $final_order->is_upe_redirect_processed() );
 	}
 
+	/**
+	 * Deposits order sets up payment intent.
+	 */
+	public function test_deposit_product_uses_setup_intents() {
```
Comment (peterwilsoncc) [inline tests/phpunit/payment-methods/test-class-wc-stripe-upe-payment-gateway.php:2250]: Failing: I think I've done the mock incorrectly.

```
There was 1 error:

1) WC_Stripe_UPE_Payment_Gateway_Test::test_deposit_product_uses_setup_intents
Cannot modify header information - headers already sent by (output started at /tmp/wordpress-tests-lib/includes/bootstrap.php:261)

/tmp/wordpress/wp-includes/pluggable.php:1450
/tmp/wordpress/wp-includes/pluggable.php:1560
/var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:1600
/var/www/html/wp-content/plugins/woocommerce-gateway-stripe/tests/phpunit/payment-methods/test-class-wc-stripe-upe-payment-gateway.php:2313
phpvfscomposer:///var/www/html/wp-content/plugins/woocommerce-gateway-stripe/vendor/phpunit/phpunit/phpunit:106
```

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php [hunk 2]
```diff
@@ -301,6 +302,16 @@ public function is_enabled_at_checkout( $order_id = null, $account_domestic_curr
...
 
+		// If cart or order contains a product that requires a payment token, enable payment method if it's reusable.
+		if ( $this->cart_contains_deposit() || ( ! empty( $order_id ) && $this->order_contains_deposit( $order_id ) ) ) {
+			return $this->is_reusable();
+		}
+
+		// If cart or order contains a product that requires a payment token, enable payment method if it's reusable.
+		if ( $this->cart_contains_deposit() || ( ! empty( $order_id ) && $this->order_contains_deposit( $order_id ) ) ) {
+			return $this->is_reusable();
+		}
```
Comment (dkotter) [inline includes/payment-methods/class-wc-stripe-upe-payment-method.php]: This seems to just be a duplicate of the above unless I'm missing something in looking at this

**Review by dkotter — CHANGES_REQUESTED** [review]

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php (hunk 2 above)
Comment (peterwilsoncc) [reply includes/payment-methods/class-wc-stripe-upe-payment-method.php]: > This seems to just be a duplicate of the above unless I'm missing something in looking at this

You're not missing anything, it is. I've removed it in d33b94d1cb85e8b633cc49f04f20c96b71972292

**Review by dkotter — APPROVED** [review]

**Comment by dkotter** [comment]
@vikrampm1 This is ready for the next step but because we normally don't work on Stripe, not sure what that step should be, if we should be aiming to QA this or just pass to the appropriate Woo team to review

**Comment by vikrampm1** [comment]
@dkotter we can do QA and then pass it over to Woo to take it from there. However, not sure how to track this over Linear as we also don't have the Stripe team setup. This might be a one-off that we need to track here and hand off once QA is complete cc @ankitguptaindia @qasumitbagthariya

**Comment by DustinHartzler** [comment]
@vikrampm1 -- will this only work with the Stripe plugin? Or can it work with WooPayments too?

**Comment by peterwilsoncc** [comment]
@DustinHartzler WooPayments will need to add the feature, similar to support for Pre-orders and Subscriptions. On the Deposits PT, I was asked to prioritize this extension.

The code for the Deposits PR is in https://github.com/woocommerce/woocommerce-deposits/pull/720 (ie, unreleased). If WooPayments will require additional hooks, please let me know.

**Comment by jeffpaul** [comment]
@ebinnion any additional reviews/testing needed from your perspective (as this extension is generally outside our scope of work / knowledge on next steps)?

**Comment by ebinnion** [comment]
@diegocurbelo - As porter, can you review the approach in this pull request? This would be following any work on 9.5.2. Let's please not merge yet though.

**Comment by qasumitbagthariya** [comment]
Hello @peterwilsoncc 

1. I couldn’t find points 11 and 12 on the [Stripe Dashboard](https://dashboard.stripe.com/):

>"Ensure that GET requests are selected in the Methods filter"

>"Ensure the payment intent detach endpoint has been hit"


2. The webhook seems to be failing. I'm testing this PR on my local system, do I need to use ngrok or InstaWP for it to work properly?

image

**Comment by peterwilsoncc** [comment]
@qasumitbagthariya Yes, webhooks need to be tested on an internet accessible server: either via ngrok or another service.

File: includes/compat/trait-wc-stripe-deposits.php [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+		if ( $top_most_order->get_payment_method() !== $this->id ) {
+			return $token;
+		}
+
+		$intent = $this->get_intent_from_order( $top_most_order );
+
+		if (
+			! $intent ||
+			! isset( $intent->setup_future_usage ) ||
+			! 'off_session' === $intent->setup_future_usage
```
Comment (malithsen) [inline includes/compat/trait-wc-stripe-deposits.php:72]: `! 'off_session' === …` parses as `(! 'off_session') === …`, which is always `false === string`

```suggestion
			! ( 'off_session' === $intent->setup_future_usage )
```

File: includes/compat/trait-wc-stripe-deposits.php [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+				$request['setup_future_usage'] = 'off_session';
+				return $request;
+			};
+
+			add_filter( 'wc_stripe_generate_create_intent_request', $future_payments );
+
+			$source   = $this->prepare_order_source( $order ); // @phpstan-ignore-line (prepare_order_source is defined in the classes that use this trait)
+			$response = $this->create_and_confirm_intent_for_off_session( $order, $source ); // @phpstan-ignore-line (create_and_confirm_intent_for_off_session is defined in the classes that use this trait)
+
+			add_filter( 'wc_stripe_generate_create_intent_request', $future_payments );
```
Comment (malithsen) [inline includes/compat/trait-wc-stripe-deposits.php:205]: This filter has already been added a few lines above. Did you mean to remove it here?

File: tests/phpunit/payment-methods/test-class-wc-stripe-upe-payment-gateway.php [hunk 5]
```diff
@@ -2241,6 +2244,97 @@ public function test_pre_order_without_payment_uses_setup_intents() {
 		$this->assertTrue( (bool) $final_order->get_meta( '_stripe_upe_redirect_processed', true ) );
 	}
 
+	/**
+	 * Deposits order sets up payment intent.
+	 */
+	public function test_deposit_product_uses_setup_intents() {
```
Comment (malithsen) [inline tests/phpunit/payment-methods/test-class-wc-stripe-upe-payment-gateway.php:2250]: Thanks for adding this test. This is not a blocker for merging but it would be great if we could add tests to cover negative paths (`order_contains_deposit`) , retries (`process_order_token_payment`), and deletion logic (`delete_order_payment_token`) as well.

**Review by malithsen — CHANGES_REQUESTED** [review]
Code-only review: I left some suggestions in comments, but the overall approach looks good to me. Thanks for working on this!

**Comment by peterwilsoncc** [comment]
Closing this off as the Deposits PT has been put on hold.
